core/run/src/lib.rs

```diff
@@ -484,3 +492,92 @@ fn execute_transactions(
 
     Ok(resp)
 }
+
+pub fn set_hardfork_info(
+    config: Config,
+    hardfork_info: Option<HardforkInfoInner>,
+) -> ProtocolResult<()> {
+    let path_rocksdb = config.data_path_for_rocksdb();
+    if !path_rocksdb.exists() {
+        let msg = format!(
+            "Data directory {} doesn't exist, please initialize it before run.",
+            path_rocksdb.display()
+        );
+        return Err(MainError::Other(msg).into());
+    }
+
+    let rt = RuntimeBuilder::new_multi_thread()
+        .enable_all()
+        .build()
+        .expect("new tokio runtime");
+
+    rt.block_on(async move {
+        log::info!("Load databases.");
+        let db_group = DatabaseGroup::new(
+            &config.rocksdb,
+            path_rocksdb,
+            false,
+            config.executor.triedb_cache_size,
+        )?;
+
+        let storage = db_group.storage();
+        let trie_db = db_group.trie_db();
+        let inner_db = db_group.inner_db();
+
+        let current_block = storage.get_latest_block(Context::new()).await?;
+        let current_state_root = current_block.header.state_root;
+
+        // Init system contract
+        let mut backend = AxonExecutorApplyAdapter::from_root(
+            current_block.header.state_root,
+            Arc::clone(&trie_db),
+            Arc::clone(&storage),
+            Proposal::new_without_state_root(&current_block.header).into(),
+        )?;
+
+        system_contract::init(inner_db, &mut backend);
+
+        let metadata_root = AxonExecutorReadOnlyAdapter::from_root(
+            current_state_root,
+            Arc::clone(&trie_db),
+            Arc::clone(&storage),
+            Proposal::new_without_state_root(&current_block.header).into(),
+        )?
+        .get_metadata_root();
+
+        let metadata_handle = MetadataHandle::new(metadata_root);
+        metadata_handle.init_hardfork(current_block.header.number)?;
+        if let Some(info) = hardfork_info {
+            if current_block.header.number >= info.block_number {
+                return Err::<(), protocol::ProtocolError>(
+                    MainError::Other(format!(
+                        "Hardfork start block number {} less than current number {}",
+                        info.block_number, current_block.header.number
+                    ))
+                    .into(),
+                );
+            }
+
+            if info.flags == H256::zero() {
+                return Ok(());
+            }
+
+            let hardforks = metadata_handle.hardfork_infos()?;
+
+            if let Some(v) = hardforks.inner.last() {
+                // if hardfork is already activated, return
+                if v.flags & info.flags == info.flags {
```

Comment:
```suggestion
                if v.flags & info.flags == v.flags {
```
Maybe it's better to rename:
- v -> current_info
- info -> proposed_info

Reply:
your opinion is wrong

There are two ways to judge:

`0b11(current) & 0b10(proposal) == 0b10`, feature already activated
`0b11 & 0b01 == 0b01`, feature already activated
`0b11 & 0b100 != 0b100`, feature can proposal

or

`0b11 | 0b10 == 0b11`,feature already activated
`0b11 | 0b01 == 0b11`,feature already activated
`0b11 | 0b100 != 0b11`, feature can proposal

Reply:
Yes, you are right.